### PR TITLE
composite-checkout: Move stripe wrapper to host page

### PR DESCRIPTION
**Pull request by sirbrillig**
#### Changes proposed in this Pull Request

This PR changes how payment methods load Stripe.js by forcing the host page to load it rather than handling the loading themselves. The motivation was so that we can check for Apple Pay availability before loading any payment methods, but it has the added advantage of simplifying the API of payment methods. 

Previously each payment method object could have a `checkoutWrapper` property which would be used to wrap the entirety of the checkout form. Although this could have other applications the intent was to make it possible for payment methods that required a shared stripe context to create one in state that was shared between sibling components in the form.

With this PR, we remove the concept of a `checkoutWrapper` and instead simply make it a requirement of stripe payment methods (currently credit card and apple pay) that the host page load, configure, initialize, and pass the `stripe` instance into the payment methods itself. This is actually quite a lot more flexible since the host page may want to load stripe in any number of ways and it prevents the possibility of loading stripe more than once on the same checkout.

With that done, this PR creates a hook `useIsApplePayAvailable` which it uses to determine if it should add the apple pay method.

One side effect of this change is that we no longer need to duplicate the stripe library code in the `composite-checkout` package; we can just use the version already on the host page (calypso). (The duplicate library still remains in the package but it is only used for the demo.)

#### Testing instructions

- Sandbox the store.
- First ignore apple pay and try visiting composite checkout in a non-Safari browser. Make sure that the Stripe payment method loads correctly.
- Using Mac OS Safari, visit WordPress.com and log in. This is required because once we proxy, you will not be able to log in.
- Enable Charles proxy to spoof WordPress.com. This is required because Apple Pay will only work on SSL and on a production domain. The details of how to do this are beyond the scope of this PR. Contact me if you have any issues.
- Using Mac OS Safari, reload WordPress.com and be sure you see the "dev" badge in the lower-right.
- Add a plan to your cart and visit checkout.
- Select "Apple Pay" from the available payment methods.
- Complete the form and press "Pay". You must use a real Apple Pay card. But as long as the store is sandboxed, you won't actually be charged.
- Complete the Apple Pay sheet that appears.
- Verify that the payment was successful and that the plan was added to your site.

#### Screenshots

Screen Shot 2020-02-25 at 12 20 46 AM

Screen Shot 2020-02-25 at 12 20 55 AM

### CODE REVIEW

**Comment by nbloomf** [comment]
Trying to load checkout in Chrome I get this error:

image

Then the form status gets stuck in the 'loading' state and visually the page hangs.

**Comment by nbloomf** [comment]
Similar behavior in Safari (which I expected) though the error is reported a little differently

image

File: packages/composite-checkout/src/lib/stripe.js [hunk 1]
```diff
@@ -242,7 +241,7 @@ function useStripeConfiguration( fetchStripeConfiguration ) {
 			debug( 'loading stripe configuration' );
 			fetchStripeConfiguration()
```
Comment (nbloomf) [inline packages/composite-checkout/src/lib/stripe.js]: Should we be passing in a `wpcom` object here?

Comment (sirbrillig) [reply packages/composite-checkout/src/lib/stripe.js]: This should be fixed 🤞

**Comment by sirbrillig** [comment]
The issues _should_ be fixed now. I had introduced some new bugs.

**Comment by nbloomf** [comment]
Still working on testing Apple Pay locally, but when I loaded composite checkout in Chrome I got this error along with "There was a problem with the submit button":

image

**Comment by sirbrillig** [comment]
Good catch. For some reason I can't get existing card payment methods to show up for me so I haven't been able to test them. Try it now.

**Comment by nbloomf** [comment]
Thanks for that update. Now that I can load checkout I'll try to test this pretty thoroughly and post bugs here as I find them.

Here's one. The steps to replicate are as follows:
1. Enter checkout with a plan upgrade
2. Select “new card” payment method, fill out the card form and proceed to the details step
3. Enter contact details

At the end of this process my payment method was no longer completed. The Name field was blank, and after filling it in again the “Continue” button did not become active, rendering the checkout uncompletable.

**Comment by nbloomf** [comment]
Here's another case, probably related to the previous one but definitely worth testing again after that bug is fixed.

1. Enter checkout with a plan.
2. Choose "new card" payment method, fill out the card details, and continue to the details step
3. Fill out billing details
4. Edit the payment method and select an existing card or paypal
5. Proceed to the payment step and submit

Here the checkout form gets stuck in a submitting state.

**Comment by nbloomf** [comment]
Here's another; not sure if this is related to the current PR. Steps to replicate are:

1. Have a site with a plan upgrade but no domain
2. Add a mapping submission via the "add a domain I already own" flow

The behavior I see is that checkout starts out in the loading state, but then redirects back to my plans page before actually loading. I still have a mapping sub in my cart, and reentering checkout from the plans page works as expected (with only the Free Purchase payment method)

**Comment by sirbrillig** [comment]
> At the end of this process my payment method was no longer completed. The Name field was blank, and after filling it in again the “Continue” button did not become active, rendering the checkout uncompletable.

Thanks for finding that. It was bad; the payment method was being recreated due to an obscure data modification. Fixed.

> Here the checkout form gets stuck in a submitting state.

I wasn't able to reproduce. Maybe it was caused by the same issue.

> Add a mapping submission via the "add a domain I already own" flow

When I tried this I was directed to regular checkout. Were you using the flag to force composite checkout to be enabled? If so, I'm not too concerned since we haven't enabled domains yet.

**Comment by nbloomf** [comment]
> Maybe it was caused by the same issue.

I'm pretty confident it was.

> Were you using the flag to force composite checkout to be enabled?

Yes; we can look at this again when testing domains.

**Comment by nbloomf** [comment]
Nice work! I was able to successfully complete payments using various combinations of

- a new card, a stored card, paypal, free product, full credits, and stored card + partial credits
- wpcom signup, wpcom upgrade, and jetpack upgrade

with no checkout related errors.

I've still had no luck getting a local apple pay environment working, but as we've discussed with the above cases tested I'm comfortable with seeing you demo that flow on your machine and then testing it live, as at this point I'm pretty confident this PR can't make apple pay _more_ broken.

**Pull request merged**